HW2/P3/mandelbrot.pyx

```diff
@@ -16,62 +24,60 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                  int max_iterations=511):
     cdef:
        int i, j, iter
-       np.complex64_t c, z
+       AVX.float8 c_real, c_imag, z_real, z_imag, counts, eight_ones, z_squared, mask, limit, eight_ones_masked, z_real_temp, z_imag_temp
+       np.ndarray[np.float32_t, ndim=2] in_coords_real, in_coords_imag,
+       float tmp_counts[8]
 
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+    # split complex numbers in real and imaginary parts
+    in_coords_real = np.real(in_coords)
+    in_coords_imag = np.imag(in_coords)
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
-                for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
-                        break
-                    z = z * z + c
-                out_counts[i, j] = iter
-
-
+        for i in prange(in_coords.shape[0], num_threads=2, schedule='static', chunksize=1):
+            for j in range(0, in_coords.shape[1], 8):
+
+                # initialize variables
+                c_real = AVX.make_float8(in_coords_real[i, j+7], in_coords_real[i, j+6],
+                                         in_coords_real[i, j+5], in_coords_real[i, j+4],
+                                         in_coords_real[i, j+3], in_coords_real[i, j+2],
+                                         in_coords_real[i, j+1], in_coords_real[i, j+0])
+
+                c_imag = AVX.make_float8(in_coords_imag[i, j+7], in_coords_imag[i, j+6],
+                                         in_coords_imag[i, j+5], in_coords_imag[i, j+4],
+                                         in_coords_imag[i, j+3], in_coords_imag[i, j+2],
+                                         in_coords_imag[i, j+1], in_coords_imag[i, j+0])
+
+                z_real = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
+                z_imag = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
+                counts = AVX.make_float8(0, 0, 0, 0, 0, 0, 0, 0)
```

Comment:
I think there is a command you can use that allows you to pass in just one number, and a float8 is created.